### PR TITLE
Possibility to replace Typed extensions via ActorSystemSetup, #24954

### PR DESCRIPTION
**Pull request by patriknw**
Refs #24954

### CODE REVIEW

File: akka-actor-typed/src/main/scala/akka/actor/typed/Extensions.scala [hunk 1]
```diff
@@ -144,3 +155,24 @@ trait Extensions {
...
+  val createExtension: java.util.function.Function[ActorSystem[_], T])
+  extends Setup
+
+/**
+ * Scala 2.11 API: Each extension typically provide a concrete `ExtensionSetup` that can be used in
+ * [[akka.actor.setup.ActorSystemSetup]] when starting the [[ActorSystem]] to replace the default
+ * implementation of the extension. Intended for tests that need to replace
+ * extension with stub/mock implementations.
+ */
+abstract class AbstractExtensionSetup[T <: Extension](extId: ExtensionId[T], createExtension: ActorSystem[_] ⇒ T)
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/Extensions.scala]: My thinking regarding the API here is that this class will not be needed when we remove support for Scala 2.11. Having overloaded function parameters doesn't work well, so the API in `ExtensionSetup` is a plain java.util.function.Function, which can be used from both Java and Scala (>= 2.12).

File: akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala [hunk 2]
```diff
@@ -38,3 +38,18 @@ class ActorRefResolver(system: ActorSystem[_]) extends Extension {
...
+
+}
+
+/**
+ * Can be used in [[akka.actor.setup.ActorSystemSetup]] when starting the [[ActorSystem]]
+ * to replace the default implementation of the [[ActorRefResolver]] extension. Intended
+ * for tests that need to replace extension with stub/mock implementations.
+ */
+final class ActorRefResolverSetup(createExtension: java.util.function.Function[ActorSystem[_], ActorRefResolver])
+  extends ExtensionSetup[ActorRefResolver](ActorRefResolver, createExtension)
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/ActorRefResolver.scala]: Nice, this may help with my serializer testkit PR!

**Review by johanandren — APPROVED** [review]
Nice!

**Review by chbatey — APPROVED** [review]
nice, lgtm

**Comment by patriknw** [comment]
fyi @hseeberger since you had some request around testing that involved typed extensions. Next step is to complete https://github.com/akka/akka/pull/24878 including defining interfaces for all typed extensions.

**Comment by hseeberger** [comment]
@patriknw will this or #24878 also help with testing behaviors which make use of the untyped system, e.g. via creating an `ActorMaterializer`?

**Comment by hseeberger** [comment]
Actually I don't want to test any interactions with the `ActorMaterializer`. What bothers me is that by just adding the creation of an `ActorMaterializer` to a behavior (`Behaviors.setup`) which formerly could be tested using the `BehaviorTestKit` leads to failing tests.

In my use case I'm testing the root actor (guardian). As the `ActorSystem` is created with the root behavior, there is no way to create an `ActorMaterializer` up front and pass it along as a parameter. Not being able to use `BehaviorTestKit` is a pity in this case, because spawning and watching can be testes very nicely with it.

**Comment by johanandren** [comment]
#24895 could make that possible together with the changes from this PR

**Comment by hseeberger** [comment]
Sounds good!

**Pull request merged**